File: King of Modules/Backups/13-03-26 v2.0.0 --- Python Zorro/Takes.py
import MetaTrader5 as mt5
 
from Config import cfg, KamaRegime, TakeMode
from Indicators import get_atr, get_kama_color
 
 
# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
 
def _resolve_kama_regime(color_code: int) -> KamaRegime:
    """Mirror of ResolveKAMARegime()."""
    if color_code == 1:
        return KamaRegime.BEARISH
    if color_code == 2:
        return KamaRegime.BULLISH
    return KamaRegime.NONE
 
 
# ---------------------------------------------------------------------------
# FIXED ATR-BASED TP — CalculateTakeProfit
# ---------------------------------------------------------------------------
 
def calculate_take_profit(
    order_type: int,   # mt5.ORDER_TYPE_BUY / ORDER_TYPE_SELL
    rates,
    symbol: str
) -> float:
    """
    Mirror of CalculateTakeProfit().
    Returns 0.0 when InpUseTakeProfit is False or ATR unavailable.
    """
    if not cfg.use_take_profit:
        return 0.0
 
    result = get_atr(rates)
    if result is None:
        return 0.0
 
    atr, _ = result
 
    info   = mt5.symbol_info(symbol)
    digits = info.digits
 
    tp_dist = atr * cfg.tp_multiplier
 
    tick = mt5.symbol_info_tick(symbol)
    if tick is None:
        return 0.0
 
    if order_type == mt5.ORDER_TYPE_BUY:
        return round(tick.ask + tp_dist, digits)
    else:
        return round(tick.bid - tp_dist, digits)
 
 
# ---------------------------------------------------------------------------
# KAMA REGIME EXIT — TakeKAMA
# ---------------------------------------------------------------------------
 
def take_kama(rates, symbol: str) -> None:
    """
    Mirror of TakeKAMA().
    BUY  position → close when KAMA flips Bullish → Bearish.
    SELL position → close when KAMA flips Bearish → Bullish.
    """
    positions = mt5.positions_get(symbol=symbol)
    if not positions:
        return
 
    result = get_kama_color(rates)
    if result is None:
        return
 
    prev_code, cur_code = result
    prev_regime = _resolve_kama_regime(prev_code)
    cur_regime  = _resolve_kama_regime(cur_code)
 
    for pos in positions:
        if pos.magic != cfg.magic_number:
            continue
 
        if pos.type == mt5.POSITION_TYPE_BUY:
            if (prev_regime == KamaRegime.BULLISH and
                    cur_regime == KamaRegime.BEARISH):
                req = {
                    "action":   mt5.TRADE_ACTION_DEAL,
                    "symbol":   symbol,
                    "volume":   pos.volume,
                    "type":     mt5.ORDER_TYPE_SELL,
                    "position": pos.ticket,
                    "price":    mt5.symbol_info_tick(symbol).bid,
                    "deviation": 10,
                    "magic":    cfg.magic_number,
                    "type_filling": mt5.ORDER_FILLING_IOC,
                }
                result_send = mt5.order_send(req)
                if result_send and result_send.retcode == mt5.TRADE_RETCODE_DONE:
                    print("TakeKAMA: BUY closed (Bullish→Bearish).")
                else:
                    code = result_send.retcode if result_send else "N/A"
                    print(f"ERROR: TakeKAMA failed to close BUY. Code={code}")
 
        elif pos.type == mt5.POSITION_TYPE_SELL:
            if (prev_regime == KamaRegime.BEARISH and
                    cur_regime == KamaRegime.BULLISH):
                req = {
                    "action":   mt5.TRADE_ACTION_DEAL,
                    "symbol":   symbol,
                    "volume":   pos.volume,
                    "type":     mt5.ORDER_TYPE_BUY,
                    "position": pos.ticket,
                    "price":    mt5.symbol_info_tick(symbol).ask,
                    "deviation": 10,
                    "magic":    cfg.magic_number,
                    "type_filling": mt5.ORDER_FILLING_IOC,
                }
                result_send = mt5.order_send(req)
                if result_send and result_send.retcode == mt5.TRADE_RETCODE_DONE:
                    print("TakeKAMA: SELL closed (Bearish→Bullish).")
                else:
                    code = result_send.retcode if result_send else "N/A"
                    print(f"ERROR: TakeKAMA failed to close SELL. Code={code}")
 
 
# ---------------------------------------------------------------------------
# ManageTakes — called every tick
# ---------------------------------------------------------------------------
 
def manage_takes(rates, symbol: str) -> None:
    """
    Mirror of ManageTakes().
    Validates magic number then dispatches to the active take mode.
    """
    positions = mt5.positions_get(symbol=symbol)
    if not positions:
        return
 
    # At least one EA-owned position must exist
    if not any(p.magic == cfg.magic_number for p in positions):
        return
 
    if cfg.take_mode == TakeMode.KAMA:
        take_kama(rates, symbol)
    # TakeMode.FIXED → no action needed here;
    # TP price level is set at entry and managed by the broker.

File: King of Modules/Zorro/Backups/19-03-2026 v1.0.0/Takes.c
// ============================================================
//  Takes.c  –  Zorro translation of Takes.mqh
//  Original: EA_KingofModules v1.30
//  Authors:  Daniel Pereira & Lucas Mattos
// ============================================================
//
//  ARCHITECTURE NOTES (MQL5 → Zorro)
//  ------------------------------------------------------------------
//  All TP functions in MQL5 returned ABSOLUTE PRICE levels.
//  Zorro's TakeProfit global expects a DISTANCE (price units).
//  GetTakeProfit() returns a DISTANCE. Callers (Entries.c) assign
//  directly to TakeProfit = GetTakeProfit(...).
//
//  ENUM_KAMA_REGIME → direct C enum translation (no changes).
//
//  ResolveKAMARegime → maps color-buffer int to regime.
//    MQL5 color code 1 → BEARISH  (KAMA falling, red)
//    MQL5 color code 2 → BULLISH  (KAMA rising, blue)
//    NOTE: color codes depend on the original "KAMA with filter"
//    indicator. In our Indicators.c re-implementation:
//      GetKAMAColor returns 0 = rising (was color 2 / BULLISH)
//                           1 = falling (was color 1 / BEARISH)
//    The mapping below is therefore ADJUSTED to the Indicators.c
//    convention:
//      0 → KAMA_REGIME_BULLISH
//      1 → KAMA_REGIME_BEARISH
//
//  take_pos.Select / g_position.PositionType → TradeIsLong /
//    TradeIsShort inside for(open_trades) loop.
//
//  g_trade.PositionClose(_Symbol) → exitTrade(ThisTrade)
//    closes the specific trade currently iterated.
//
//  ManageTakes: original checked magic number then called TakeKAMA.
//    In Zorro the script owns all trades; no magic filter needed.
//    Runs for every open trade; direction tested inside TakeKAMA.
//
//  CPositionInfo take_pos → removed.
// ============================================================

#ifndef TAKES_C
#define TAKES_C

#include "Indicators.c"


// ============================================================
//  ENUM_KAMA_REGIME
// ============================================================
typedef enum {
    KAMA_REGIME_NONE    = 0,
    KAMA_REGIME_BEARISH = 1,
    KAMA_REGIME_BULLISH = 2
} ENUM_KAMA_REGIME;


// ============================================================
//  ResolveKAMARegime
//  Maps the integer color code from GetKAMAColor() to a regime.
//
//  Indicators.c convention (re-implemented KAMA):
//    0 → KAMA rising  → BULLISH
//    1 → KAMA falling → BEARISH
//
//  Original MQL5 "KAMA with filter" used:
//    1 → BEARISH (red),  2 → BULLISH (blue)
//  Mapping is adjusted accordingly.
// ============================================================
ENUM_KAMA_REGIME ResolveKAMARegime(int colorCode)
{
    if(colorCode == 0) return KAMA_REGIME_BULLISH;
    if(colorCode == 1) return KAMA_REGIME_BEARISH;
    return KAMA_REGIME_NONE;
}


// ============================================================
//  CalculateTakeProfit  (internal)
//  Returns a TP DISTANCE (positive, price units).
//
//  MQL5 original returned absolute price; Zorro needs distance.
//  Distance = ATR * InpTPMultiplier + extra points buffer.
// ============================================================
static var CalculateTakeProfit_ATR()
{
    var atrVal = 0, atrAvg = 0;
    if(!GetATR(&atrVal, &atrAvg))
        return 0.0;

    var extraDist = InpTakeExtraPoints * POINT_VALUE;
    var tpDist    = atrVal * InpTPMultiplier + extraDist;

    return tpDist;
}


// ============================================================
//  GetTakeProfit  (public – called from Entries.c)
//  Returns a TP DISTANCE (positive, price units).
//
//  TAKE_FIXED  → ATR-based distance (same as TAKE_KAMA at entry;
//                KAMA exit is managed bar-by-bar in ManageTakes)
//  TAKE_POINTS → fixed distance in points (InpTakePoints)
//  TAKE_KAMA   → ATR-based initial distance; KAMA flip closes it
// ============================================================
var GetTakeProfit(ENUM_SIGNAL signal, var atr)
{
    if(!InpUseTakeProfit)
        return 0.0;

    ENUM_TAKE_MODE mode = (signal == SIGNAL_BUY) ? InpLongTakeMode
                                                  : InpShortTakeMode;

    if(mode == TAKE_POINTS)
        return InpTakePoints * POINT_VALUE;

    // TAKE_FIXED and TAKE_KAMA: ATR-based distance at entry
    var extraDist = InpTakeExtraPoints * POINT_VALUE;
    return atr * InpTPMultiplier + extraDist;
}


// ============================================================
//  TakeKAMA  (internal per-trade)
//  Called inside for(open_trades) – ThisTrade is current trade.
//
//  Closes the trade when KAMA regime flips against the position:
//    LONG  exit: previous bar BULLISH → current bar BEARISH
//    SHORT exit: previous bar BEARISH → current bar BULLISH
// ============================================================
static void TakeKAMA()
{
    int prev = 0, cur = 0;
    if(!GetKAMAColor(&prev, &cur))
        return;

    ENUM_KAMA_REGIME prevRegime = ResolveKAMARegime(prev);
    ENUM_KAMA_REGIME curRegime  = ResolveKAMARegime(cur);

    if(TradeIsLong)
    {
        if(prevRegime == KAMA_REGIME_BULLISH && curRegime == KAMA_REGIME_BEARISH)
        {
            exitTrade(ThisTrade);
            printf("TakeKAMA: LONG closed – KAMA flipped Bullish→Bearish");
        }
    }
    else if(TradeIsShort)
    {
        if(prevRegime == KAMA_REGIME_BEARISH && curRegime == KAMA_REGIME_BULLISH)
        {
            exitTrade(ThisTrade);
            printf("TakeKAMA: SHORT closed – KAMA flipped Bearish→Bullish");
        }
    }
}


// ============================================================
//  ManageTakes  (public – called from EA_KingofModules.c)
//  Runs every bar for all open trades (no time filter).
//  Each trade is evaluated and closed individually if the
//  KAMA exit condition is met.
//
//  Only TAKE_KAMA mode triggers active management here.
//  TAKE_FIXED and TAKE_POINTS rely on the TakeProfit price level
//  set at entry time (handled by Zorro's broker sync).
// ============================================================
void ManageTakes()
{
    if(!InpUseTakeProfit)
        return;

    // Only KAMA mode requires bar-by-bar management
    if(InpLongTakeMode  != TAKE_KAMA &&
       InpShortTakeMode != TAKE_KAMA)
        return;

    for(open_trades)
    {
        // Apply KAMA take only to trades whose direction matches
        // the configured mode
        if(TradeIsLong  && InpLongTakeMode  == TAKE_KAMA) TakeKAMA();
        if(TradeIsShort && InpShortTakeMode == TAKE_KAMA) TakeKAMA();
    }
}

#endif // TAKES_C
